### Neste código avaliamos a precisão numerica da linguagem de python para o calculo de simulações fisicas

In [1]:
def f(x):
    return x*(x-1)

x=1
delta = 10**(-2)

print((f(x+delta)-f(x))/delta)

1.010000000000001


O valor encontrado analítico foi da derivada sendo igual a 1 e o numérico foi de 1.010000000000001. O erro do valor foi portanto de 0.010000000000001 significando que nossa aproximação tem um erro muito grande se comparado ao valor analítico. Dessa foma seria importante diminuirmos o valor do delta para melhorar a precisão. Esse erro ocorre por causa da imprecisão numérica existente na divisão de um numero <code>float</code>.

In [3]:
x=1
delta = [10**(-4), 10**(-6), 10**(-8), 10**(-10), 10**(-12), 10**(-14)]
resultados = []

for d in delta: 
    resultados.append((f(x+d)-f(x))/d)

print(resultados)

[1.0000999999998899, 1.0000009999177333, 1.0000000039225287, 1.000000082840371, 1.0000889005833413, 0.9992007221626509]


Nessa ultima celula podemos ver o esperado, o valor da aproximação fica mais proximo do real até o valor de $10^{-8}$ e depois o erro volta a aumentar. Isso ocorre por causa da imprecisão inata do numero em <code>float</code> no Python, que tem precisão somente até esse valor de potência.